File: dme_loader/src/jenkins.cpp
#include "jenkins.h"
#include <cstring>

namespace jenkins {
    uint32_t lookup2(const char* key, uint32_t init) {
        size_t length = std::strlen(key);
        size_t position = length;
        std::shared_ptr<uint32_t> a(new uint32_t), b(new uint32_t), c(new uint32_t);
        uint32_t p = 0;
        *a = 0x9e3779b9;
        *b = *a;
        *c = init;
        while(position >= 12) {
            *a += key[p+0] + (((uint32_t)key[p+1]) << 8) + (((uint32_t)key[p+2]) << 16) + (((uint32_t)key[p+3]) << 24);
            *b += key[p+4] + (((uint32_t)key[p+5]) << 8) + (((uint32_t)key[p+6]) << 16) + (((uint32_t)key[p+7]) << 24);
            *c += key[p+8] + (((uint32_t)key[p+9]) << 8) + (((uint32_t)key[p+10]) << 16) + (((uint32_t)key[p+11]) << 24);
            mix(a, b, c);
            p += 12;
            position -= 12;
        }
        *c += length;
        if(position >= 11) *c += ((uint32_t)key[p+10])<<24;
        if(position >= 10) *c += ((uint32_t)key[p+9])<<16;
        if(position >= 9) *c += ((uint32_t)key[p+8])<<8;
        if(position >= 8) *b += ((uint32_t)key[p+7])<<24;
        if(position >= 7) *b += ((uint32_t)key[p+6])<<16;
        if(position >= 6) *b += ((uint32_t)key[p+5])<<8;
        if(position >= 5) *b += ((uint32_t)key[p+4]);
        if(position >= 4) *a += ((uint32_t)key[p+3])<<24;
        if(position >= 3) *a += ((uint32_t)key[p+2])<<16;
        if(position >= 2) *a += ((uint32_t)key[p+1])<<8;
        if(position >= 1) *a += ((uint32_t)key[p+0]);

        mix(a, b, c);
        return *c;
    }

    void mix(std::shared_ptr<uint32_t> a, std::shared_ptr<uint32_t> b, std::shared_ptr<uint32_t> c) {
        *a -= *b;
        *a -= *c;
        *a ^= (*c) >> 13;

        *b -= *c;
        *b -= *a;
        *b ^= (*a) << 8;

        *c -= *a;
        *c -= *b;
        *c ^= (*b) >> 13;

        *a -= *b;
        *a -= *c;
        *a ^= (*c) >> 12;

        *b -= *c;
        *b -= *a;
        *b ^= (*a) << 16;
        
        *c -= *a;
        *c -= *b;
        *c ^= (*b) >> 5;
        
        *a -= *b;
        *a -= *c;
        *a ^= (*c) >> 3;
        
        *b -= *c;
        *b -= *a;
        *b ^= (*a) << 10;
        
        *c -= *a;
        *c -= *b;
        *c ^= (*b) >> 15;
    }

    // Note: Forgelight uses a signed 32 bit int when hashing, but returns it as unsigned
    uint32_t oaat(const char* key) {
        size_t length = strlen(key);
        size_t i = 0;
        int32_t hash = 0;
        while(i < length) {
            hash += (signed char)key[i];
            hash += hash << 10;
            hash ^= hash >> 6;
            i++;
        }
        hash += hash << 3;
        hash ^= hash >> 11;
        hash += hash << 15;
        return (uint32_t)hash;
    }
}

File: script/jenkins.py
from typing import Tuple
from numpy import int32, uint32, int8
import warnings


# Note: Forgelight uses a signed 32 bit int while performing the Jenkins hash, but represents it as unsigned
def oaat(key: bytes):
    with warnings.catch_warnings():
        warnings.filterwarnings("ignore", "overflow encountered in int_scalars", RuntimeWarning)
        warnings.filterwarnings("ignore", "overflow encountered in long_scalars", RuntimeWarning)
        i = 0
        hash = int32(0)
        while i < len(key):
            hash += int8(key[i])
            i += 1
            hash += int32(hash << 10)
            hash ^= int32(hash >> 6)
        hash += int32(hash << 3)
        hash ^= int32(hash >> 11)
        hash += int32(hash << 15)
        return int(uint32(hash))

def mix(a: int, b: int, c: int) -> Tuple[int, int, int]:
    a &= 0xFFFFFFFF
    b &= 0xFFFFFFFF
    c &= 0xFFFFFFFF

    a -= b
    a -= c
    a ^= c >> 13
    a &= 0xFFFFFFFF

    b -= c
    b -= a
    b ^= a << 8
    b &= 0xFFFFFFFF

    c -= a
    c -= b
    c ^= b >> 13
    c &= 0xFFFFFFFF

    a -= b
    a -= c
    a ^= c >> 12
    a &= 0xFFFFFFFF

    b -= c
    b -= a
    b ^= a << 16
    b &= 0xFFFFFFFF

    c -= a
    c -= b
    c ^= b >> 5
    c &= 0xFFFFFFFF

    a -= b
    a -= c
    a ^= c >> 3
    a &= 0xFFFFFFFF

    b -= c
    b -= a
    b ^= a << 10
    b &= 0xFFFFFFFF

    c -= a
    c -= b
    c ^= b >> 15
    c &= 0xFFFFFFFF

    return a, b, c

def lookup2(key: bytes, initval: int = 0):
    lenpos = len(key)
    length = lenpos
    a = b = 0x9e3779b9
    c = initval
    p = 0

    while lenpos >= 12:
        a += key[p+0] + key[p+1] << 8 + key[p+2] << 16 + key[p+3] << 24
        a &= 0xFFFFFFFF
        b += key[p+4] + key[p+5] << 8 + key[p+6] << 16 + key[p+7] << 24
        b &= 0xFFFFFFFF
        c += key[p+8] + key[p+9] << 8 + key[p+10] << 16 + key[p+11] << 24
        c &= 0xFFFFFFFF
        a, b, c = mix(a, b, c)
        p += 12
        lenpos -= 12
    
    c += length
    if (lenpos >= 11): c += key[p+10]<<24
    if (lenpos >= 10): c += key[p+9]<<16
    if (lenpos >= 9):  c += key[p+8]<<8
    # the first byte of c is reserved for the length
    if (lenpos >= 8):  b += key[p+7]<<24
    if (lenpos >= 7):  b += key[p+6]<<16
    if (lenpos >= 6):  b += key[p+5]<<8
    if (lenpos >= 5):  b += key[p+4]
    if (lenpos >= 4):  a += key[p+3]<<24
    if (lenpos >= 3):  a += key[p+2]<<16
    if (lenpos >= 2):  a += key[p+1]<<8
    if (lenpos >= 1):  a += key[p+0]

    a, b, c = mix(a, b, c)

    return c & 0xFFFFFFFF